junit-vintage-engine/src/main/java/org/junit/vintage/engine/VintageTestEngine.java

```diff
@@ -75,11 +88,64 @@ public void execute(ExecutionRequest request) {
 
 	private void executeAllChildren(VintageEngineDescriptor engineDescriptor,
 			EngineExecutionListener engineExecutionListener) {
+		boolean parallelExecutionEnabled = getParallelExecutionEnabled();
+
+		if (parallelExecutionEnabled) {
+			executeInParallel(engineDescriptor, engineExecutionListener);
+		}
+		else {
+			executeSequentially(engineDescriptor, engineExecutionListener);
+		}
+	}
+
+	private void executeInParallel(VintageEngineDescriptor engineDescriptor,
+			EngineExecutionListener engineExecutionListener) {
+		ExecutorService executorService = Executors.newFixedThreadPool(getThreadPoolSize());
+		RunnerExecutor runnerExecutor = new RunnerExecutor(engineExecutionListener);
+
+		List<TestDescriptor> children = new CopyOnWriteArrayList<>(engineDescriptor.getModifiableChildren());
+
+		CompletableFuture<Void> allOf = CompletableFuture.allOf(
+			children.stream().map(descriptor -> CompletableFuture.runAsync(() -> {
+				RunnerTestDescriptor testDescriptor = (RunnerTestDescriptor) descriptor;
+				try {
+					runnerExecutor.execute(testDescriptor);
+				}
+				catch (Exception e) {
+					engineExecutionListener.executionSkipped(testDescriptor, e.getMessage());
+				}
+			}, executorService)).toArray(CompletableFuture[]::new));
+
+		try {
+			allOf.get(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+		}
+		catch (Exception e) {
+			logger.warn(e, () -> "Timeout or interruption while waiting for parallel test execution to finish");
+			Thread.currentThread().interrupt();
```

Comment:
We should only do this for `InterruptedException`. `ExecutionExceptions` should be unwrapped and rethrown using `ExceptionUtils.throwAsUncheckedException`.